File: v8_liskov_substitution_principle/java/Main1.java
package v8_liskov_substitution_principle.java;

public class Main1 {

    public static void main(String[] args) {
        Calculator c1 = new Calculator();
        System.out.println(c1.calculate(5, 10));
        SuperCalculator c2 = new SuperCalculator();
        System.out.println(c2.calculate(5, 10));
    }
    
}

class Calculator {
    public int calculate(int n1, int n2) {
        return n1 + n2;
    }
}

class SuperCalculator {
    public int calculate(int n1, int n2) {
        return n1 - n2;
    }
}

File: v8_liskov_substitution_principle/python/main1.py
import abc

class Calculator:

    @abc.abstractmethod
    def test():
        raise NotImplementedError
    
    def calculate(self, n1, n2):
        return n1 + n2

class SuperCalculator(Calculator):

    def test():
        print('super test')

    def calculate(self, n1, n2): # It's dangerous to overwrite non-abstract methods
        return n1 - n2

c = Calculator()
print(c.calculate(1, 3))
sc = SuperCalculator()
print(sc.calculate(1, 3))
